firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java

```diff
@@ -191,12 +192,23 @@ private String readJsonString(int bytesToRead) throws IOException {
    */
   private boolean pullMoreData() throws IOException {
     buffer.compact();
-    int bytesToRead = Math.min(bundleInputStream.available(), buffer.remaining());
+
+    int available = bundleInputStream.available();
+    int bytesToRead = Math.min(available, buffer.remaining());
+    // `available` is an estimation, we still try to read if the estimation is 0, to move things
+    // forward.
+    if (available == 0) {
+      bytesToRead = buffer.remaining();
+    }
     byte[] bytes = new byte[bytesToRead];
     int bytesRead = bundleInputStream.read(bytes);
-    buffer.put(bytes);
+    boolean readSuccess = bytesRead > 0;
+    if (readSuccess) {
+      buffer.put(bytes, 0, bytesRead);
```

Comment:
Could you make the call to `buffer.put(bytes, 0, bytesRead)` unconditional? Or is specifying `bytesRead==0` going to cause problems?

Reply:
`bytesRead` can be -1, that is the problem.

Reply:
Ahh I see. Makes sense.